util/postprocess-makedepend.pl

```diff
@@ -0,0 +1,160 @@
+            # won't get mixed up with the target / deps delimiter.  We use this
+            # fact in the regexp below to make sure we do look at the target.
+            $line =~ s/^/$directory/ unless /^\S+[:>\]]\S+\s+:/;
+
+            # We know that VMS has system header files in text libraries,
+            # extension .TLB.  We also know that our header files aren't stored
+            # in text libraries.  Finally, we know that VMS C produces exactly
+            # one dependency per line, so we simply discard any line ending with
+            # .TLB.
+            return undef if /\.TLB\s*$/;
+
+            my ($target, $deps) = $line =~ /^(.*)\s:\s(.*)/;
+            $deps =~ s/^\s+//;
+            $deps =~ s/\s+$//;
+            return ($target, $deps);
+        },
+    'VC' =>
+        sub {
+            my $line = shift;
+            my $object = shift;
+
+            # For the moment, we only support Visual C on native Windows, or
+            # compatible compilers.  With those, the flags /Zs /showIncludes
+            # give us the necessary output to be able to create dependencies
+            # that nmake (or any 'make' implementation) should be able to read,
+            # with a bit of help.  The output we're interested in looks like
+            # this (it always starts the same)
+            #
+            #   Note: including file: {whatever header file}
+            #
+            # So all we really have to do is to is to replace the start of the
+            # line with an object file specification, given to us as an extra
+            # argument (passed from $ARGV[1]);
+            #
+            # There are also other lines mixed in, for example compiler
+            # warnings, so we simply discard anything that doesn't start with
+            # the Note:
+
+            if (/^Note: including file: */) {
+                (my $tail = $') =~ s/\s*\R$//;
+
+                # VC gives us absolute paths for all include files, so to
+                # remove system header dependencies, we need to check that
+                # they don't match $abs_srcdir or $abs_blddir
+                $tail = canonpath($tail);
+                if ($tail =~ m|^\Q$abs_srcdir\E|i
+                        || $tail =~ m|^\Q$abs_blddir\E|i) {
+                    return ($object, "\"$tail\"");
+                }
+            }
+
+            return undef;
+        },
+} -> {$producer};
+
+die "Producer unrecognised: $producer\n" unless defined $procedure;
+
+my %collect = ();
+while (<STDIN>) {
+    s|\R$||;                    # The better chomp
+    my ($target, $deps) = $procedure->($_, @ARGV);
+    $collect{$target}->{$deps} = 1
+        if defined $target;
+}
+
+my $continuation = {
+    'makedepend' => "\\",
+    'VMS C' => "-",
+    'VC' => "\\",
+} -> {$producer};
+
+die "Producer unrecognised: $producer\n" unless defined $continuation;
+
+foreach my $target (sort keys %collect) {
```

Comment:
How many keys can be in %collect? I mean isn't meaning that there will be just one? But even if there can be more than one (on VMS?), wouldn't its value being a list instead of hash be more efficient? That list doesn't have to be sorted, because compiler will always return it in same order...

Reply:
> How many keys can be in %collect? I mean isn't meaning that there will be just one?

That's right now, yeah, but I'm trying to make this script generic enough so that if we change our minds on that at some point, it can already handle more than one target.  Boom.

> wouldn't its value being a list instead of hash be more efficient?

Well...  see this for example:

> ```
> : ; makedepend -f- -o'|apps/opt.o' -- -I ../master/ -I ../master/include/ -I include/ -- ../master/apps/opt.c 2>/dev/null
> # DO NOT DELETE
> 
> ../master/apps/opt|apps/opt.o: ../master/apps/apps.h ../master/e_os.h
> ../master/apps/opt|apps/opt.o: /usr/include/limits.h
> ../master/apps/opt|apps/opt.o: /usr/include/bits/libc-header-start.h
> ../master/apps/opt|apps/opt.o: /usr/include/features.h
> ../master/apps/opt|apps/opt.o: /usr/include/stdc-predef.h
> ../master/apps/opt|apps/opt.o: /usr/include/sys/cdefs.h
> ../master/apps/opt|apps/opt.o: /usr/include/bits/wordsize.h
> ...
> ../master/apps/opt|apps/opt.o: ../master/include/openssl/engineerr.h
> ../master/apps/opt|apps/opt.o: ../master/include/openssl/ocsp.h
> ../master/apps/opt|apps/opt.o: ../master/include/openssl/x509v3.h
> ../master/apps/opt|apps/opt.o: ../master/include/openssl/x509v3err.h
> ../master/apps/opt|apps/opt.o: ../master/include/openssl/ocsperr.h
> ...
> ```

If we collect that in a list, we gotta handle repeated targets yada yada yada...  much simpler to use a hash, which isn't *that* inefficient.

Reply:
Oh, did you mean that I made a hash of the dependencies as well?  What if any of them is repeated, wouldn't you want to shave that off as well?

Reply:
> What if any of them is repeated

Fair enough.